fdbserver/IPager.h

```diff
@@ -91,75 +233,350 @@ class ArenaPage : public ReferenceCounted<ArenaPage>, public FastAllocated<Arena
+
+		// Populate encryption key with relevant fields from page
+		if (page->encodingType == EncodingType::XOREncryption) {
+			encryptionKey.id = page->getEncodingHeader<XOREncryptionEncodingHeader>()->keyID;
+		}
 
-	void updateChecksum(LogicalPageID pageID) { getChecksum() = calculateChecksum(pageID); }
+		if (page->headerVersion == 1) {
+			if (verify) {
+				RedwoodHeaderV1* h = page->getMainHeader<RedwoodHeaderV1>();
+				h->verifyChecksum(buffer, pPayload - buffer);
+				if (pageID != h->firstPhysicalPageID) {
+					throw page_header_wrong_page_id();
+				}
+			}
+		} else {
+			throw page_header_version_not_supported();
+		}
+	}
 
-	bool verifyChecksum(LogicalPageID pageID) { return getChecksum() == calculateChecksum(pageID); }
+	// Pre:   postReadHeader has been called, encoding-specific parameters (such as the encryption secret) have been set
+	// Post:  Payload has been verified and decrypted if necessary
+	void postReadPayload(PhysicalPageID pageID) {
+		if (page->encodingType == EncodingType::XXHash64) {
+			page->getEncodingHeader<XXHashEncodingHeader>()->decode(pPayload, payloadSize, pageID);
+		} else if (page->encodingType == EncodingType::XOREncryption) {
+			ASSERT(encryptionKey.secret.size() == 1);
+			page->getEncodingHeader<XOREncryptionEncodingHeader>()->decode(
+			    encryptionKey.secret[0], pPayload, payloadSize, pageID);
+		} else {
+			throw page_encoding_not_supported();
+		}
+	}
 
 	const Arena& getArena() const { return arena; }
 
+	static bool isEncodingTypeEncrypted(EncodingType t) { return t != EncodingType::XXHash64; }
+
+	// Returns true if the page's encoding type employs encryption
+	bool isEncrypted() const { return isEncodingTypeEncrypted(getEncodingType()); }
+
 private:
 	Arena arena;
+
+	// The logical size of the page, which can be smaller than bufferSize, which is only of
+	// practical purpose in simulation to use arbitrarily small page sizes to test edge cases
+	// with shorter execution time
 	int logicalSize;
+
+	// The 4k-aligned physical size of allocated memory for the page which also represents the
+	// block size to be written to disk
 	int bufferSize;
-	uint8_t* buffer;
+
+	// buffer is a pointer to the page's memory
+	// For convenience, it is unioned with a Page pointer which defines the page structure
+	union {
+		uint8_t* buffer;
+		const Page* page;
+	};
+
+	// Pointer and length of page space available to the user
+	// These are accessed very often so they are stored directly
+	uint8_t* pPayload;
+	int payloadSize;
 
 public:
-	mutable void* userData;
-	mutable void (*userDataDestructor)(void*);
+	EncodingType getEncodingType() const { return page->encodingType; }
+
+	PhysicalPageID getPhysicalPageID() const {
+		if (page->headerVersion == 1) {
+			return page->getMainHeader<RedwoodHeaderV1>()->firstPhysicalPageID;
+		} else {
+			throw page_header_version_not_supported();
+		}
+	}
+
+	// Used by encodings that do encryption
+	EncryptionKey encryptionKey;
```

Comment:
Encyrption key caching should be done using BlobCipherKeyCache utility class (Refer to BlobCipher.h for details); the class provides freebees such as: EK derivation from baseCipher, EK TTL, avoid logging plaintext EK as part of core-dumps etc.

Reply:
We can certainly change this later.  Some sort of weak-reference to a thing owned and managed by the BlobCipherKeyCache would be good, so we aren't doing a "lookup" for each page written necessarily as we can check to see if the weak references is still valid first.

Reply:
Reference<BlobCipherKey> seems like the right thing to use here (in-memory representation of encryption-key for FDB codebase, it is used to encrypt buffer using EncryptBlobCipherAes265Ctr class and has enough context to allow generation of encryption header (baseCipherId, randomSalt etc.) , BlobCipherKeyCache is a library and every FDB encryption participant should instantiate an object for its usage (singleton). On cache miss the caller process is supposed to have logic to reach out to EncryptKeyProxy endpoint to refresh/fetch the encryption keys (code documents have details for library behavior, if you find them insufficient, please let me know and I should improve on it)

Reply:
I haven't looked at that code yet, but what I was thinking is lookup() returns a reference to an object which you can hold long term and get the key from it but that key may be removed by the library at any time, so then when you go to use it you have to do a lookup again.  This prevents some kind of map lookup into a cache container for each cached key usage but also avoids any code outside the library having a copy of the secret not under direct control by the library.  I'm calling this a "weak reference" because while you have a reference to an object with the secret, your holding of that object doesn't keep the secret present, the library will remove it when it wants to.